cpp/oneapi/dal/algo/dbscan/backend/gpu/kernel_fp_impl.hpp

```diff
@@ -60,40 +60,44 @@ struct get_core_wide_kernel {
+            std::int64_t wg_size = get_recommended_wg_size(queue, column_count);
             cgh.parallel_for(
                 bk::make_multiple_nd_range_2d({ wg_size, block_size }, { wg_size, 1 }),
                 [=](sycl::nd_item<2> item) {
-                    auto sg = item.get_sub_group();
-                    const std::uint32_t sg_id = sg.get_group_id()[0];
-                    if (sg_id > 0)
-                        return;
-                    const std::uint32_t wg_id = item.get_global_id(1);
-                    if (wg_id >= block_size)
-                        return;
-                    const std::uint32_t local_id = sg.get_local_id();
-                    const std::uint32_t local_size = sg.get_local_range()[0];
+                    auto sg = item.get_group();
+
+                    const std::uint32_t row_id = item.get_global_id(1);
+
+                    const std::uint32_t local_id = item.get_local_id(0);
 
                     count_type count = 0;
                     for (std::int64_t j = 0; j < row_count; j++) {
                         Float sum = Float(0);
-                        for (std::int64_t i = local_id; i < column_count; i += local_size) {
-                            Float val = data_ptr[(block_start + wg_id) * column_count + i] -
+                        Float distance = Float(0);
+                        for (std::int64_t i = local_id; i < column_count; i += wg_size) {
+                            Float val = data_ptr[(block_start + row_id) * column_count + i] -
                                         data_ptr[j * column_count + i];
                             sum += val * val;
+
+                            distance +=
+                                sycl::reduce_over_group(sg, sum, sycl::ext::oneapi::plus<Float>());
+                            if (distance > epsilon) {
+                                break;
+                            }
                         }
-                        Float distance =
-                            sycl::reduce_over_group(sg, sum, sycl::ext::oneapi::plus<Float>());
-                        if constexpr (use_weights) {
-                            count += distance <= epsilon ? weights_ptr[j] : count_type(0);
-                        }
-                        else {
-                            count += distance <= epsilon ? count_type(1) : count_type(0);
+                        if (distance <= epsilon) {
+                            count += use_weights ? weights_ptr[j] : count_type(1);
+                            if (count >= min_observations) {
+                                if (local_id == 0) {
+                                    cores_ptr[row_id] = count_type(1);
+                                }
+                                break;
+                            }
+                            if (row_count - j + count < min_observations) {
+                                break;
+                            }
```

Comment:
I think we can apply this break only if use_weights is set to False. If it is set to true then potentially an instance could be a core sample if the sum of w_j,...,w_count plus count is greater than min_observation. Considering that weights could be more than 1 using this if statement we may miss some core camples.